File: SkyEditor.Core/IO/ISavableAs.vb
Namespace IO

    ''' <summary>
    ''' Implemented by objects that can save themselves to disk.
    ''' </summary>
    Public Interface ISavableAs
        Inherits ISavable
        ''' <summary>
        ''' Saves the class to the given filename.
        ''' </summary>
        ''' <param name="Filename"></param>
        Overloads Sub Save(Filename As String, provider As IOProvider)

        ''' <summary>
        ''' Gets the default extension for the file.
        ''' </summary>
        ''' <returns>A string containing the default extension for the file.</returns>
        Function GetDefaultExtension() As String

        ''' <summary>
        ''' Gets the supported extensions for the file.
        ''' </summary>
        ''' <returns>An IEnumerable that contains every extension that can be used to save this file.</returns>
        Function GetSupportedExtensions() As IEnumerable(Of String)
    End Interface

End Namespace

File: SkyEditor.Core/IO/ISavableAs.cs
using SkyEditor.IO.FileSystem;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace SkyEditor.Core.IO
{
    /// <summary>
    /// Represents a class that can save itself to disk at a specific filename
    /// </summary>
    public interface ISavableAs : ISavable
    {
        /// <summary>
        /// Saves to the given filename
        /// </summary>
        /// <param name="filename">Path of the destination file</param>
        /// <param name="provider">Instance of the current I/O provider</param>
        Task Save(string filename, IFileSystem provider);

        /// <summary>
        /// Gets the default extension for the file
        /// </summary>
        /// <returns>The default extension of the file</returns>
        string GetDefaultExtension();

        /// <summary>
        /// Gets the supported extensions for the file
        /// </summary>
        /// <returns>The supported extensions that can be used to save the file</returns>
        IEnumerable<string> GetSupportedExtensions();
    }
}
